theano/tensor/subtensor.py

```diff
@@ -2001,7 +2001,7 @@ def perform(self, node, inp, out_):
         if self.set_instead_of_inc:
             x[idx] = y
         else:
-            if config.cxx:
+            if config.cxx and node.inputs[0].dtype != 'float16':
```

Comment:
I don't think it really is an issue with numpy, more that `float16` is not a type in C, just an alias for `int16`.
So the "increment" calls "+" on int16, and then we interpret the result as float16, which does not make sense.

A better way of implementing that from Theano's end could be to use `ufunc.at`, rather than loading a precompiled module. I just opened #5904.